mimxrt: Add initial PSRAM implementation.

sfe-SparkFro opened the pull request:
### Summary

PSRAM support for Teensy 4.1 and other `mimxrt` boards! Resolves #18281

This implementation is a copy of the [Arduino core PSRAM implementation here](https://github.com/PaulStoffregen/cores/blob/a23cade55c7530cce43c641ec3ddb8b7f83148a8/teensy4/startup.c#L379-L587). I modified the implementation to use the `#define`s from [`lib/nxp_driver/sdk/devices/MIMXRT1062/MIMXRT1062.h`](https://github.com/micropython/nxp_driver/blob/master/sdk/devices/MIMXRT1062/MIMXRT1062.h) where possible instead of the huge [`imxrt.h`](https://github.com/PaulStoffregen/cores/blob/master/teensy4/imxrt.h). However there's still about a dozen that I had to copy over from `imxrt.h` that currently live in `psram.c`; happy to change/move these if there's a better solution!

The MicroPython side of things is a copy of the `rp2` PSRAM PR (#15620). I have not looked into whether `MICROPY_ALLOC_GC_STACK_SIZE` needs to be increased nor if any clock timings need to change, like was done in the `rp2` PR. There may be other edge cases that need to be addressed, but I'm new to Teensy/mimxrt, so don't know what to look for :slightly_smiling_face:

### Testing

<!-- Explain what testing you did, and on which boards/ports. If there are
     boards or ports that you couldn't test, please mention this here as well.

     If you leave this empty then your Pull Request may be closed. -->

The following test code prints the total amount of memory seen by the garbage collector, then performs a memory test with a small and large `bytearray` to demonstrate the split heap working.

```
import gc
import uctypes

gc.collect()
print("Initial GC free (MiB):", gc.mem_free() / 1024 / 1024)

def test(size):
    gc.collect()
    b = bytearray(size)
    print("Test size:", size, "location: 0x{:X}".format(uctypes.addressof(b)), end=' ')
    for i in range(size):
        b[i] = i & 0xFF
    for i in range(size):
        if b[i] != (i & 0xFF):
            print("Fail at", i)
            return
    print("Pass!")

test(10)
test(10_000_000)
```

Testing with a single 16MiB chip, the GC shows over 16MiB available. The small array is allocated in the `0x20000000` range (DTCM), and the large array is allocated in the `0x70000000` range (PSRAM). All behaves as I'd expect!

```
Initial GC free (MiB): 16.358551025390626
Test size: 10 location: 0x20204720 Pass!
Test size: 10000000 location: 0x7005DD00 Pass!
```

Note that `test(10_000_000)` takes ~25 seconds on my end.

### Trade-offs and Alternatives

<!-- If the Pull Request has some negative impact (i.e. increased code size)
     then please explain why you think the trade-off improvement is worth it.
     If you can think of alternative ways to do this, please explain that here too.

     Delete this heading if not relevant (i.e. small fixes) -->

There could be an argument for PSRAM support being enabled by default, at least for Teensy 4.1 (set `#define MICROPY_HW_ENABLE_PSRAM (1)` in `boards/TEENSY41/mpconfigboard.h`). This implementation auto-detects the presence and size of 1 or 2 PSRAM chips if they're soldered, meaning a user wouldn't need to flash different firmware.

Building on my end, enabling PSRAM increases the firmware size from 626992 to 628064 bytes (+1072 / +0.17%), so I'll defer to others whether it's worth enabling it by default. If not, might be nice to create a variant with PSRAM enabled so it's easy to get from the [Teensy 4.1 download page](https://micropython.org/download/TEENSY41/).

sfe-SparkFro committed:
mimxrt: Add initial PSRAM implementation.

Signed-off-by: Dryw Wade <dryw.wade@sparkfun.com>

sfe-SparkFro commented:
Oh also, this would need permission from @PaulStoffregen, the code I copied does not have a license. Would be ideal if it could have the MIT license or similar added!

robert-hh commented:
You can enable PSRAM by default for Teensy 4.1. If the chip is not present (or blocked), just internal RAM is used. Tested on my board.
About the formatting errors. You can fix them with uncrustify using e.g.
`uncrustify --no-backup -c ../../tools/uncrustify.cfg psram.c`
You may have to install uncrustify.

robert-hh commented on ports/mimxrt/psram.c:
```diff
@@ -0,0 +1,258 @@
+
+// Copied from here:
+// https://github.com/PaulStoffregen/cores/blob/a23cade55c7530cce43c641ec3ddb8b7f83148a8/teensy4/avr/pgmspace.h#L31
+#define FLASHMEM __attribute__((section(".flashmem")))
```
There is no reference to .flashmem in the loader file. So this definition seems obsolete. All code goes to flash anyhow, unless told otherwise.

sfe-SparkFro replied:
Done in [8b2a3be](https://github.com/micropython/micropython/pull/18288/commits/8b2a3be18772fdaffebed4fa12dbed024b1c00db)

robert-hh commented on ports/mimxrt/psram.c:
```diff
@@ -0,0 +1,258 @@
+#define FLEXSPI_LUTKEY_VALUE			((uint32_t)0x5AF05AF0)
+#define FLEXSPI_LUT_INSTRUCTION(opcode, pads, operand) ((uint32_t)(\
+	(((opcode) & 0x3F) << 10) | (((pads) & 0x03) << 8) | ((operand) & 0xFF)))
+#define FLEXSPI_LUT_OPCODE_CMD_SDR		0x01
```
There is already flexspi_flash_config.h which defines these constants and the LUT entry constructor, maybe worded slightly different.  So maybe it's better to switch to these definitions. It's in the context of flash, but the definitions are for flexspi. If the name of the file matters, it could be unified.

sfe-SparkFro commented:
> Works on my Teensy 4.1 with an 8 MB RAM chip.

Great to hear!

> I found once a note by Paul Stoffregen telling that the PJRC code is licensed under MIT or MIT-Like conditions: https://forum.pjrc.com/index.php?threads/open-source-license-issues-when-using-teensy-products.29382/. But that statement is pretty old.

Paul just pushed https://github.com/PaulStoffregen/cores/commit/10025393e83ca9f4dc5646643a41cb2f32022ae4 which adds license header comments.

> You can enable PSRAM by default for Teensy 4.1.

> About the formatting errors. You can fix them with uncrustify

Done!

sfe-SparkFro committed:
mimxrt: Improve PSRAM implementation.

Enable PSRAM by default for Teensy 4.1,
update license info, uncrustify psram.c
and main.c, and remove FLASHMEM macro.

Signed-off-by: Dryw Wade <dryw.wade@sparkfun.com>

robert-hh commented on ports/mimxrt/psram.c:
```diff
@@ -0,0 +1,237 @@
+// https://github.com/PaulStoffregen/cores/blob/10025393e83ca9f4dc5646643a41cb2f32022ae4/teensy4/imxrt.h
+#define CCM_CCGR_ON                             3
+#define FLEXSPI_MCR2_CLRLEARNPHASE(x)           ((uint32_t)(x << 14))
+#define FLEXSPI_LUTKEY_VALUE                    ((uint32_t)0x5AF05AF0)
```
All of these are defined in fsl_flexspi.h as well as :

```
#define FREQ_1MHz             (1000000UL)
#define FLEXSPI_DLLCR_DEFAULT (0x100UL)
#define FLEXSPI_LUT_KEY_VAL   (0x5AF05AF0UL)
```

robert-hh commented on ports/mimxrt/psram.c:
```diff
@@ -0,0 +1,237 @@
+    IOMUXC->SELECT_INPUT_1[kIOMUXC_FLEXSPI2_IPP_IND_IO_FA_BIT3_SELECT_INPUT] = 1;     // GPIO_EMC_29 for Mode: ALT8
+    IOMUXC->SELECT_INPUT_1[kIOMUXC_FLEXSPI2_IPP_IND_SCK_FA_SELECT_INPUT] = 1;     // GPIO_EMC_25 for Mode: ALT8
+
+    // turn on clock  (QSPI flash & PSRAM chips usually spec max clock 100 to 133 MHz)
```
The clock setup in the following lines down to  line 137 inc. can be made a little bit more readable using the NXP lib macros:
```
    // Set FLEXSPI2_PODF divide by 5.
    CLOCK_SetDiv(kCLOCK_Flexspi2Div, 4);
    // Set Flexspi2 clock source to PLL2, 528 MHz.
    CLOCK_SetMux(kCLOCK_Flexspi2Mux, 3);
    CLOCK_EnableClock(kCLOCK_FlexSpi2);
```
The various other combinations may be added as a comment:

robert-hh commented on ports/mimxrt/psram.c:
```diff
@@ -0,0 +1,237 @@
+
+// These #defines are from here:
+// https://github.com/PaulStoffregen/cores/blob/10025393e83ca9f4dc5646643a41cb2f32022ae4/teensy4/imxrt.h
+#define CCM_CCGR_ON                             3
```
This #define is not used.

robert-hh commented:
There is DCACHE and ICACHE enabled in the port, which must be disabled when writing to flash. I do not know where the regions are defined.

sfe-SparkFro committed:
mimxrt/psram: Improve PSRAM implementation.

Add missing bullet in license.
Use symbolic names for pins.
Use macros for setting FlexSPI2 clock.

Signed-off-by: Dryw Wade <dryw.wade@sparkfun.com>

robert-hh commented:
Looks better. If it works it's right. With the call to `CLOCK_SetDiv()` I had use symbolic values for the second argument, which reflect the effective divider, so instead of 
`CLOCK_SetDiv(kCLOCK_Flexspi2Div, 5)` it is `CLOCK_SetDiv(kCLOCK_Flexspi2Div, kCLOCK_Flexspi2DivBy6)`. kCLOCK_Flexspi2DivBy6 is defined as `(5)`, etc....
There might be a way to use the lib functions for configuring the flexspi unit instead of direct register access of lines 153ff. But even if possible that would require setting up a sparsely documented data structure with the intended configuration, and that may be not better readable than the existing code.

sfe-SparkFro commented:
> If it works it's right.

Yes, still works fine on my end!

> instead of `CLOCK_SetDiv(kCLOCK_Flexspi2Div, 5)` it is `CLOCK_SetDiv(kCLOCK_Flexspi2Div, kCLOCK_Flexspi2DivBy6)`.

I see those are defined in [`_clock_div_value`](https://github.com/micropython/nxp_driver/blob/91b04b34a59f6d81661cec6f84611afe6330ce92/sdk/devices/MIMXRT1062/drivers/fsl_clock.h#L839-L846), happy to swap that out!

Do you know if there's an equivalent enum for `CLOCK_SetMux()` to replace the `3`? I don't see a `_clock_mux_value` enum or anything.


The original implementation uses `CCM_CBCMR_FLEXSPI2_CLK_SEL(3)`, which according to [`MIMXRT1062.h`](https://github.com/micropython/nxp_driver/blob/master/sdk/devices/MIMXRT1062/MIMXRT1062.h) should be `derive clock from PLL2 (pll2_main_clk)`:

```c
/*! FLEXSPI2_CLK_SEL
 *  0b00..derive clock from PLL2 PFD2
 *  0b01..derive clock from PLL3 PFD0
 *  0b10..derive clock from PLL3 PFD1
 *  0b11..derive clock from PLL2 (pll2_main_clk)
 */
#define CCM_CBCMR_FLEXSPI2_CLK_SEL(x)            (((uint32_t)(((uint32_t)(x)) << CCM_CBCMR_FLEXSPI2_CLK_SEL_SHIFT)) & CCM_CBCMR_FLEXSPI2_CLK_SEL_MASK)
```

So I thought maybe [`kCLOCK_SysPllClk`](https://github.com/micropython/nxp_driver/blob/91b04b34a59f6d81661cec6f84611afe6330ce92/sdk/devices/MIMXRT1062/drivers/fsl_clock.h#L491), but that has a value of `0xE0`.

Note that [`CLOCK_SetMux`](https://github.com/micropython/nxp_driver/blob/91b04b34a59f6d81661cec6f84611afe6330ce92/sdk/devices/MIMXRT1062/drivers/fsl_clock.h#L1408-L1426) does the exact same bit shifting (once you work through the nested macros), so a value of `3` is required as the second parameter:

```c
static inline void CLOCK_SetMux(clock_mux_t mux, uint32_t value)
{
    uint32_t busyShift;

    busyShift               = (uint32_t)CCM_TUPLE_BUSY_SHIFT(mux);
    CCM_TUPLE_REG(CCM, mux) = (CCM_TUPLE_REG(CCM, mux) & (~CCM_TUPLE_MASK(mux))) |
                              (((uint32_t)((value) << CCM_TUPLE_SHIFT(mux))) & CCM_TUPLE_MASK(mux));
...
}
```

So should I just leave it as `CLOCK_SetMux(kCLOCK_Flexspi2Mux, 3)`?

> There might be a way to use the lib functions for configuring the flexspi unit instead of direct register access

Looks like the correct functions are [here](https://github.com/micropython/nxp_driver/blob/master/sdk/drivers/flexspi/fsl_flexspi.h) (eg. `FLEXSPI_Init()`). Will see about changing that to see if it looks more readable!

sfe-SparkFro committed:
mimxrt/psram: Use lib functions in favor of direct register access.

Signed-off-by: Dryw Wade <dryw.wade@sparkfun.com>

sfe-SparkFro committed:
mimxrt/psram: Change to normal MIT license.
